[Student learingn] Fixed Fix droplet toolbox for rtl language appearance

levadadenys opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->
### [Student learingn] Fixed Fix droplet toolbox for rtl language appearance
“Show/hide toolbox” in droplet in RTL languages is on the right side of the header, but the toolbox is still on the left. This menu should stay above the toolbox.
Before:
![Screenshot from 2023-02-01 19-09-02](https://user-images.githubusercontent.com/22244040/216114598-0c27dd6a-1d1f-4567-bc17-a3d539ebdd13.png)

![Screenshot from 2023-02-01 19-08-58](https://user-images.githubusercontent.com/22244040/216114115-8a68b758-1623-4320-a027-587ad5d7f101.png)

After:
![Screenshot from 2023-02-02 21-38-16](https://user-images.githubusercontent.com/22244040/216433079-140612f4-0df9-40bd-8c8a-dae61ae1f5b3.png)
![Screenshot from 2023-02-02 21-38-21](https://user-images.githubusercontent.com/22244040/216433097-01b538d4-1bc6-4540-aa7b-f552c447e7ff.png)



## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

mikeharv commented:
This is interesting. When I read the issue, I assumed that the task was just to reorder what was in the purple header so that it matched what is below. Mock-up:
![Untitled](https://user-images.githubusercontent.com/43474485/216117521-f4fac690-341f-49ae-983e-6746e1b239ad.jpg)

I could be wrong! 

However, if we do want to move the toolbox, I'm noticing that right edge of the workspace and the left edge of the toolbox seem to blend together. I wonder if we could also move the code mirror gutter to the right of the workspace (and directly to the left of the toolbox). Mock-up:

![image](https://user-images.githubusercontent.com/43474485/216118368-19f36a7c-5ee6-4b25-96af-b1c1d77c6788.png)


 Alternatively, I think we'd want to add some kind of vertical divider.

![image](https://user-images.githubusercontent.com/43474485/216118513-1e0dbd0a-bd2d-4c49-8163-57e308276342.png)

levadadenys commented:
@mikeharv 
Good point, now that I've read the ticket one more time I also think maybe toolbox is meant to stay on the left. I'm gonna raise that question in jira and come back once we have an answer on that.

 thanks a lot!

levadadenys commented:
@molly-moen @jmkulwik

jmkulwik commented:
> @molly-moen @jmkulwik

I agree with Mike on this one. 

The ideal behavior is what we have in Sprite Lab [expanded toolbox](https://studio.code.org/s/coursec-2022/lessons/6/levels/7?lang=en-US), [collapsed toolbox](https://studio.code.org/projects/spritelab/new).  Anything that gets us closer to that behavior while still being a positive user experience is a good thing. It may not be very realistic to move the numbers to the right at this point, but some sort of vertical divider is needed to make the separation between the toolbox and the workspace more obvious.

pablo-code-org commented on apps/src/StudioApp.js:
```diff
@@ -1524,15 +1524,21 @@ StudioApp.prototype.resizeVisualization = function(width) {
-    var categories = document.querySelector('.droplet-palette-wrapper');
-    toolboxWidth = categories.getBoundingClientRect().width;
+  if (this.editCode && this.editor && this.editor.session) {
+    const isRtl = getStore().getState().isRtl;
+    const categories = document.querySelector('.droplet-palette-wrapper');
+
+    if (this.editor.session.paletteEnabled) {
+      // If in the droplet editor, set toolboxWidth based on the block palette width:
+      toolboxWidth = categories.getBoundingClientRect().width;
+    }
+
+    if (isRtl) {
+      // If Rtl - handle show/hide toolbox functionality
+      categories.style.zIndex = this.editor.session.paletteEnabled
+        ? 'inherit'
+        : '0';
+    }
```
nit: variables should be initialized when are going to be closed to be used
```
    if (this.editor.session.paletteEnabled) {
      // If in the droplet editor, set toolboxWidth based on the block palette width:
      toolboxWidth = categories.getBoundingClientRect().width;
    }
   const isRtl = getStore().getState().isRtl;
    const categories = document.querySelector('.droplet-palette-wrapper');

    if (isRtl) {
      // If Rtl - handle show/hide toolbox functionality
      categories.style.zIndex = this.editor.session.paletteEnabled
        ? 'inherit'
        : '0';
    } 
```

levadadenys replied:
@pablo-code-org 
1.  const categories - We need it in the first If statement, so they can't be declared later
2. const rtl - is more of a holy war object I guess. I prefer to keep all the variables declarations in the beginning of the function/block scope so that code stays easy to read and understand

levadadenys commented:
I've updated the PR and updated the screenshots of final results.
I've added a border-left to the toolbox as suggested by @mikeharv 
@mikeharv @jmkulwik feel free to review/approve.
I'll try to move numbers to the right tomorrow, but in case it won't help I thinking of merging this solution.

molly-moen reviewed (approved):
LGTM!

levadadenys commented:
@mikeharv @jmkulwik @molly-moen 
I've created a separate branch for moving gutter (numbers) to the right. [link](https://github.com/code-dot-org/code-dot-org/tree/denys/sl-428-gutter/move-gutter-to-the-right)
It moves gutter to the right when it's in static position. However, when you click show code/blocks - there's an animation(transition of gutter(numbers) on the right side) that I don't know how to update right now. If anyone'll be working on this later - I suggest starting from that branch.
Also there're some calculation to be made for handling show/hide toolbox.

levadadenys commented:
@pablo-code-org I've made a small update (changed the order of the If statements) so that it should not raise a holy war re const isRtl. Pls take a look
